doc/doxygen: add build system doc page for BOARD, CPU, FEATURE (#12473)
RIOT-OS/RIOT: doc/doxygen/src/board-cpu-features.md

FEATURES                                                             {#features}
========

> `FEATURES_PROVIDED` are hardware (including toolchain) features (e.g.:`periph_hwrng`,

Reviewer: I somehow precise it with `BSP board support package` to include also build system, build system configuration, flasher/tooling and not only `toolchain` that could be misinterpreted as `gcc/ld` only.

> `FEATURES_PROVIDED` are hardware (including toolchain) features (e.g.:`periph_hwrng`,

Reviewer 2: ```suggestion
`FEATURES_PROVIDED` are available hardware (including toolchain) features (e.g.:`periph_hwrng`,
```

> `periph_uart`) that are available or characteristics (e.g:`arch_32bits`) of a board.

Reviewer 2: ```suggestion
`periph_uart`) or characteristics (e.g:`arch_32bits`) of a board/cpu.
```

Providing a `FEATURE`

> --------------------

Reviewer 2: ```suggestion
---------------------
```

For a `FEATURE` to be provided by a `board` it must meet 2 criteria, and for

> `periph` and other `hw` it must follow the 3rd criteria.

Reviewer 2: What is `hw` ?

> `periph` and other `hw` it must follow the 3rd criteria.

Reviewer 2: ```suggestion
`periph` and other `hw` it must follow a 3rd criteria.
```

- Needs the "hardware" or BSP support (toolchain)
    - e.g.: `stm32l152re` has an SPI peripheral
            `riotboot` needs to be able to link and flash at an offset
- Needs support in RIOT, an implementation of an api to interact with the `hw`
    - e.g.: `cpu/stm32_common/periph/spi.c` is implemented for `stm32l1`
            `riotboot` needs an implementation of `cpu_jump_to_image`

> - Wiring between the `cpu/soc` a `bus` and other `cpu`/`hw` components.

Reviewer: For this one `riotboot` could also be included as it shows a non directly hardware thing but needs BSP/build system "wiring" with the usage of an adapted linker script, RIOT specific flasher characteristics, slots configuration https://github.com/RIOT-OS/RIOT/blob/15de138674652b3d83fa75a01d01efa91052aa33/cpu/cortexm_common/Makefile.include#L26-L71

Author: Mentioned on the first point!

    - e.g.: `nucleo-l152re/include/periph_conf.h` specified wiring between `PORT_Ax`
             and `SPI1`

What is a `FEATURE`

> --------------------

Reviewer 2: ```suggestion
-------------------
```

> Whenever a `FEATURE` is used there should be at some level a HW requirements,

Reviewer 2: ```suggestion
Whenever a `FEATURE` is used there should be at some level a hardware requirement,
```

> whether this is a `radio`, a `bus` ot a specific core architecture. 

Reviewer 2: ```suggestion
whether this is a `radio`, a `bus` of a specific core architecture. 
```

This is not a hard line, in some cases the line can be harder to establish than
others. There are complicated cases like `netif` since a network interface could
be fully implemented in software as a loop-back.

It's also important to note that a `FEATURE` does not mean there is a `MODULE`
with the same name. There could be many implementations for the same `FEATURE`.
The fact that many `FEATURES` translate directly into a `MODULE` is only by
convenience.

e.g.

    # all periph features correspond to a periph submodule
    USEMODULE += $(filter periph_%,$(FEATURES_USED))

> A `FEATURE` is a tool to specify valid/invalid dependencies and configurations.

Reviewer: This could be said first. Its only goal is doing automatic detection of invalid configurations and so handle one application for all RIOT boards without explicit per board configuration step. Otherwise you just add modules and know what you are doing.

It is also currently the way to do "optional inclusion of support". Adding a short info on optional features could be good. The optional thingy is as the "default implementation" a way to configure the resolution policy more than the dependency. Include by default all what is optional.

A future step could be to never show `features` at the application level but represent them with a virtual module, and add optional module dependency as we talked with @aabadie 
It is more an implementation issue to have them there and leaks this inconsistency to the user.
"But what if my feature is provided only by software, why should it be set in the board".
It is just the current state of RIOT. If the feature can be provided only by software, then all boards having this implementation must say it. But then how to configure which implementation to choose ? and this is mostly done in a non configurable way by the magic `periph_` handling.

> A `FEATURE` is a tool to specify valid/invalid dependencies and configurations.

Reviewer 2: ```suggestion
A `FEATURE` is a mean to specify valid/invalid dependencies and configurations.
```
 (or way or whatever but not tool)

CPU/CPU_MODEL                                                             {#cpu}

> ======

Reviewer 2: ```suggestion
=============
```

> `CPU` and `CPU_MODEL` refer to the `SOC` (system on a chip) or `MCU` (microcontroller)

Reviewer 2: SOC was already used earlier but not defined. I would move the definition when it's used the first time. Also no need to "highlight" it with backticks. Same for MCU. Maybe use italic ?

> present in a `BOARD`. The variables `CPU`, `CPU_FAM`, etc. are just arbitrary groupings

Reviewer 2: ```suggestion
present in a `BOARD`. The variables `CPU`, `CPU_FAM`, etc. are just arbitrary groups
```

to avoid code duplication. How this grouping is done depends on every implementation
and the way each manufacturer groups there products.

These variables allows declaring the `FEATURES` that the mcu/soc provides as well

> resolving dependencies.

Reviewer 2: ```suggestion
as resolving dependencies.
```

`FEATURES` provided by a `CPU/CPU_MODEL` should not depend on the wiring of a
specific `BOARD` but be intrinsic to the soc/mcu.

A `CPU/CPU_MODEL` might support `FEATURES` that will depend the `BOARD` wiring,
e.g.: bus (`uart`, `spi`) mappings. In this cases the `FEATURE` should be provided

> by the `BOARD.`

Reviewer: Good to precise this one. It is a tricky one to get by default and really shows that the 3 criteria must be there. 
The silicon might have it, there may be common code, but not be wired on the board or just not yet described in a `periph_conf`.

BOARDS                                                                 {#boards}
======

> In RIOT build-system terms a `BOARD` is a grouping of:

Reviewer 2: ```suggestion
In the RIOT build-system, the variable `BOARD` is a grouping of:
```

Author: I don't agree with the need of mentioning variable.

Reviewer 2: ok, but add a comma after `terms`, this is more readable.

- soc/mcu (`CPU/CPU_MODEL`)
    - e.g.: `b-l072z-lrwan1` `stm32l072cz`
- sensor/actuators (buttons and leds included) (`drivers`)
    - e.g.: `b-l072z-lrwan1` leds and buttons
- radios, ethernet, etc. devices (`drivers`)
    - e.g.: `b-l072z-lrwan1` `sx1276`
- programming/debugging tools
    - e.g.: `b-l072z-lrwan1` `stlink`
- configuration mapping cpu support capabilities to availability
    - e.g.: `b-l072z-lrwan1` `periph_conf.h`, `gpio_params`

A `board` can have all the required `FEATURES` to interact with a radio or
sensor/actuator, but it doesn't necessarily provide that `FEATURE`.

e.g.:

>     - `samr21-xpro` provides a `at86r233` radio as well as the necessary

Reviewer 2: ```suggestion
    - `samr21-xpro` provides a `at86rf233` radio as well as the necessary
```

       `periph_*` features.
    - `nucleo-*` provide all `periph_*` features to use `sx1272`, and
       even a default configuration for the `SX1272MB2xA` shield, but not
       doesn't provide the radio.

If a `board` in `$(RIOTBASE)/boards` is connected to a radio shield, sensors,

> actuators, etc. then it is a different `board` then the one provided by default.

Reviewer: ```suggestion
actuators, etc. then it is a different `board` than the one provided by default.
```
